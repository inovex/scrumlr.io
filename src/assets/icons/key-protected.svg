<svg class="icon" width="26" height="26" viewBox="0 0 26 26" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M9.87502 9.00002C9.87502 5.065 13.065 1.87502 17 1.87502C20.9351 1.87502 24.125 5.065 24.125 9.00002C24.125 12.8762 21.058 15.1866 18.3562 15.6369C17.0406 15.8562 15.71 15.6268 14.8913 15.2174C14.5393 15.0414 14.1253 15.125 13.867 15.3956L12.1742 17.0884C12.1525 17.1101 12.1097 17.1348 12.0445 17.138C11.7213 17.1535 11.2443 17.1889 10.7881 17.273C10.5604 17.315 10.3162 17.3729 10.0893 17.456C9.87714 17.5336 9.60203 17.6606 9.38131 17.8813C9.08709 18.1755 8.94492 18.5864 8.8637 18.8625C8.77453 19.1657 8.70565 19.5102 8.64616 19.8077L8.64615 19.8078L8.64202 19.8284C8.57745 20.1512 8.52245 20.4228 8.45745 20.6438C8.42323 20.7601 8.39518 20.829 8.37899 20.8652C8.3589 20.8716 8.33016 20.8795 8.29076 20.8876C8.1466 20.9176 7.95803 20.9328 7.71029 20.9441C7.66265 20.9463 7.61133 20.9483 7.5576 20.9505L7.55751 20.9505C7.3637 20.9582 7.13846 20.9671 6.94085 20.9877C6.71042 21.0116 6.34163 21.0637 6.03026 21.2618C5.65957 21.4977 5.49209 21.8544 5.41056 22.1167C5.32907 22.3788 5.30039 22.6542 5.28396 22.8633C5.27523 22.9745 5.26994 23.0617 5.26523 23.1394L5.26523 23.1395L5.26523 23.1395C5.25904 23.2417 5.25383 23.3276 5.24309 23.43C5.23135 23.5418 5.21862 23.5945 5.21475 23.6105C5.21385 23.6143 5.21343 23.616 5.2136 23.6161L5.21398 23.6156C5.14318 23.7582 5.08786 23.9001 5.05124 23.994L5.05123 23.9941L5.03455 24.0367C5.0206 24.072 5.00977 24.0984 5.00057 24.1196C4.98906 24.1215 4.97551 24.1233 4.95974 24.125H1.87502V20.9142C1.87502 20.8811 1.88819 20.8493 1.91164 20.8258L10.1045 12.633C10.2438 12.5001 10.3335 12.3259 10.3638 12.1402C10.3841 12.0156 10.3777 11.8859 10.3416 11.7598C10.3252 11.7024 10.3029 11.6463 10.2746 11.5926C9.88762 10.8081 9.87502 10.0151 9.87502 9.00002ZM8.46634 11.7963C8.12455 10.816 8.1248 9.86433 8.12501 9.05945L8.12502 9.00002C8.12502 4.0985 12.0985 0.125024 17 0.125024C21.9016 0.125024 25.875 4.0985 25.875 9.00002C25.875 13.9604 21.942 16.8134 18.6439 17.3631C17.2582 17.5941 15.8252 17.4428 14.7017 17.0358L13.4116 18.3258C13.0583 18.6792 12.5931 18.8636 12.1285 18.8859C11.8383 18.8999 11.4501 18.9305 11.1054 18.994C10.9327 19.0258 10.7925 19.0621 10.6907 19.0994C10.662 19.1099 10.6405 19.119 10.625 19.1261C10.6097 19.1594 10.5798 19.2298 10.5426 19.3563C10.4776 19.5773 10.4226 19.8488 10.358 20.1716L10.3539 20.1923L10.3539 20.1923C10.2944 20.4899 10.2255 20.8343 10.1363 21.1375C10.0551 21.4137 9.91296 21.8245 9.61874 22.1187C9.3146 22.4229 8.92689 22.5428 8.64679 22.601C8.35345 22.662 8.04202 22.6808 7.78976 22.6923C7.71711 22.6956 7.6497 22.6983 7.58647 22.7009L7.5856 22.7009H7.58553C7.40504 22.7082 7.2586 22.7141 7.1217 22.7283C7.09829 22.7307 7.07744 22.7332 7.05897 22.7357C7.04782 22.8 7.03758 22.8859 7.02859 23.0004C7.02486 23.0479 7.0208 23.1113 7.01629 23.1819C7.00745 23.32 6.99686 23.4857 6.98353 23.6127C6.96236 23.8143 6.92008 24.1165 6.78265 24.3913C6.74885 24.4589 6.7245 24.521 6.68694 24.6167L6.68693 24.6167L6.66205 24.6799C6.61757 24.7925 6.54426 24.976 6.42528 25.1527C6.29376 25.3479 6.11118 25.5304 5.85474 25.6637C5.61046 25.7907 5.34337 25.8501 5.07057 25.8722L5.03535 25.875H5.00002H1.50002C0.740633 25.875 0.125024 25.2594 0.125024 24.5V20.9142C0.125024 20.417 0.322568 19.94 0.674199 19.5884L8.46634 11.7963ZM7.08962 22.6141L7.08817 22.6174C7.08909 22.6151 7.08959 22.6141 7.08962 22.6141Z" fill="currentColor"/>
<circle cx="19" cy="7" r="2.35" stroke="currentColor" stroke-opacity="0.7" stroke-width="1.3"/>
</svg>
