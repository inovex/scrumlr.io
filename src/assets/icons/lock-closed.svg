<svg class="icon" width="32" height="32" viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M22 12H10C8.34315 12 7 13.3431 7 15V24C7 25.6569 8.34315 27 10 27H22C23.6569 27 25 25.6569 25 24V15C25 13.3431 23.6569 12 22 12Z" stroke="currentColor" stroke-width="1.75"/>
<path d="M20.56 11.7V9.55096C20.545 8.35297 20.0594 7.20896 19.2078 6.36614C18.3563 5.52332 17.2074 5.04939 16.0093 5.04676C14.8112 5.04414 13.6602 5.51302 12.805 6.3521C11.9498 7.19117 11.4591 8.33305 11.439 9.53096V11.7" stroke="currentColor" stroke-width="1.75" stroke-miterlimit="10" stroke-linecap="round" stroke-linejoin="round"/>
<path d="M16 17.652V21.352" stroke="currentColor" stroke-opacity="0.7" stroke-width="1.5" stroke-miterlimit="10" stroke-linecap="round" stroke-linejoin="round"/>
</svg>
