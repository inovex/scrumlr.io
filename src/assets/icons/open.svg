<svg class="icon" width="32" height="32" viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg">
<circle cx="16" cy="16" r="12.125" stroke="currentColor" stroke-width="1.75"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M20.4838 16.9C20.3785 19.8059 19.7695 22.3582 18.8904 24.1897C17.8697 26.3162 16.7503 27 16 27C15.2497 27 14.1303 26.3162 13.1096 24.1897C12.2305 22.3582 11.6215 19.8059 11.5162 16.9H20.4838ZM21.9847 16.9H28V15.4H21.9932C21.8429 8.77508 19.2172 3.5 16 3.5C12.7828 3.5 10.1571 8.77508 10.0068 15.4H4V16.9H10.0153C10.2367 23.3835 12.8316 28.5 16 28.5C19.1684 28.5 21.7633 23.3835 21.9847 16.9ZM11.5072 15.4C11.58 12.3699 12.2004 9.70454 13.1096 7.81026C14.1303 5.68385 15.2497 5 16 5C16.7503 5 17.8697 5.68385 18.8904 7.81026C19.7996 9.70454 20.42 12.3699 20.4928 15.4H11.5072Z" fill="currentColor" fill-opacity="0.7"/>
</svg>
