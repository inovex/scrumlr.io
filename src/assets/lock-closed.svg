<?xml version="1.0" encoding="iso-8859-1"?>
<!-- Generator: Adobe Illustrator 19.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Capa_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 477.867 477.867" style="enable-background:new 0 0 477.867 477.867;" xml:space="preserve">
<g>
	<g>
		<path d="M409.6,170.667h-17.067V153.6c0-84.831-68.769-153.6-153.6-153.6s-153.6,68.769-153.6,153.6v17.067H68.267
			c-9.426,0-17.067,7.641-17.067,17.067V460.8c0,9.426,7.641,17.067,17.067,17.067H409.6c9.426,0,17.067-7.641,17.067-17.067
			V187.733C426.667,178.308,419.026,170.667,409.6,170.667z M287.205,324.055c-5.155,14.581-16.624,26.05-31.205,31.205v37.274
			c0,9.426-7.641,17.067-17.067,17.067s-17.067-7.641-17.067-17.067V355.26c-26.66-9.426-40.631-38.679-31.205-65.339
			S229.34,249.29,256,258.716C282.66,268.142,296.631,297.395,287.205,324.055z M358.4,170.667H119.467V153.6
			c0-65.98,53.487-119.467,119.467-119.467S358.4,87.62,358.4,153.6V170.667z"/>
	</g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
<g>
</g>
</svg>
