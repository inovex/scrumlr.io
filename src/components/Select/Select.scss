@import "constants/style";

.select {
  display: flex;
  flex-direction: column;
  gap: $spacing--base;
}
