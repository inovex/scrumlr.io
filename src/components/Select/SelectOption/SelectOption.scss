@import "constants/style";

.select-option {
  display: grid;
  grid-template-rows: 2px auto auto;
  grid-template-columns: auto $spacing--sm 1fr $spacing--sm auto;
  grid-template-areas:
    "icon . .           . radio"
    "icon . label       . radio"
    ".    . description . . ";

  padding: $spacing--base;

  border: 2px solid transparent;
  border-radius: 12px;

  background-color: $gray--100;

  cursor: pointer;

  &--active {
    border-color: $blue--500;
  }

  &--extra {
    grid-template-rows: 2px auto auto $spacing--sm auto;
    grid-template-columns: auto $spacing--sm 1fr $spacing--sm auto;
    grid-template-areas:
      "icon . .           .     radio"
      "icon . label       .     radio"
      ".    . description .     .    "
      ".    . .           .     .    "
      ".    . extra       extra extra";
  }
}

.select-option__icon-container {
  grid-area: icon;

  width: $icon--large;
  height: $icon--large;
}

.select-option__label {
  grid-area: label;

  font-size: $text--base;
  font-weight: 600;
  line-height: normal;

  color: $navy--400;
}

.select-option__description {
  grid-area: description;

  font-size: $text--base;
  font-weight: 400;
  line-height: 22px;

  color: $navy--500;
}

.select-option__radio {
  grid-area: radio;

  width: 20px;
  height: 20px;
  margin: 0;
  padding: 6px;
}

.select-option__extra-content {
  grid-area: extra;
}

[theme="dark"] {
  .select-option {
    background: $navy--500;

    &--active {
      border-color: $blue--400;
    }
  }

  .select-option__label,
  .select-option__description,
  .select-option__icon-container {
    color: $gray--000;
  }
}
