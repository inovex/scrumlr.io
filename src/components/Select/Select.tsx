import {Children, cloneElement, isValidElement, ReactElement, useMemo} from "react";
import {SelectOptionProps} from "components/Select/SelectOption/SelectOption";
import {SelectContext} from "utils/hooks/useSelect";
import "./Select.scss";

type SelectProps = {
  children: ReactElement<SelectOptionProps> | ReactElement<SelectOptionProps>[];
  activeIndex: number;
  setActiveIndex: (index: number) => void;
};

// general select component
// add <SelectOption> elements as children
export const Select = ({children, activeIndex, setActiveIndex}: SelectProps) => {
  // memoize the context value to avoid unnecessary re-renders
  const contextValue = useMemo(() => ({activeIndex, setActiveIndex}), [activeIndex, setActiveIndex]);

  return (
    <SelectContext.Provider value={contextValue}>
      <div className="select">{Children.map(children, (child, index) => (isValidElement<SelectOptionProps>(child) ? cloneElement(child, {index}) : null))}</div>
    </SelectContext.Provider>
  );
};
