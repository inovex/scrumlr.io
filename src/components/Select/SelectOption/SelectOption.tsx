import {ReactNode} from "react";
import {useSelect} from "utils/hooks/useSelect";
import classNames from "classnames";
import "./SelectOption.scss";

export type SelectOptionProps = {
  label: string;
  description: string;
  icon: ReactNode;
  index?: number; // overridden by <Select> component
  children?: ReactNode; // optional extra content
};

export const SelectOption = (props: SelectOptionProps) => {
  if (props.index === undefined) throw new Error("undefined index, make sure to wrap <SelectOption> components inside a <Select> component or <SelectContext.Provider>");

  const {activeIndex, setActiveIndex} = useSelect();
  const isActive = activeIndex === props.index;

  const handleSelect = () => {
    if (props.index !== undefined) {
      setActiveIndex(props.index);
    }
  };

  const renderExtraContent = () => (props.children && isActive ? <div className="select-option__extra-content">{props.children}</div> : null);

  return (
    <button
      className={classNames("select-option", `select-option--${props.index}`, {"select-option--active": isActive}, {"select-option--extra": props.children})}
      onClick={handleSelect}
      tabIndex={0}
    >
      <div className="select-option__icon-container">{props.icon}</div>
      <div className="select-option__label">{props.label}</div>
      <div className="select-option__description">{props.description}</div>

      <input className="select-option__radio" type="radio" checked={isActive} onChange={handleSelect} />

      {renderExtraContent()}
    </button>
  );
};
