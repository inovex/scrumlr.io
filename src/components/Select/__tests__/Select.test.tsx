import {render} from "testUtils";
import {Select} from "components/Select/Select";
import {SelectOption} from "components/Select/SelectOption/SelectOption";
import {ReactComponent as KeyIcon} from "assets/icons/key-protected.svg";
import {ReactComponent as GlobeIcon} from "assets/icons/open.svg";
import {ReactComponent as LockIcon} from "assets/icons/lock-closed.svg";

const renderSelect = () =>
  render(
    <Select activeIndex={0} setActiveIndex={jest.fn}>
      <SelectOption label={"test 1"} description={"another test 1"} icon={<GlobeIcon />} />
      <SelectOption label={"test 2"} description={"another test 2"} icon={<LockIcon />} />
      <SelectOption label={"test 3"} description={"another test 3"} icon={<KeyIcon />} />
    </Select>
  );

describe("Select", () => {
  it("should render correctly", () => {
    const {container} = renderSelect();

    expect(container).toMatchSnapshot();
  });
});
