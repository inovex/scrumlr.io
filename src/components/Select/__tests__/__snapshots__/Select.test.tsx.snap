// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render correctly 1`] = `
<div>
  <div
    class="select"
  >
    <button
      class="select-option select-option--0 select-option--active"
      tabindex="0"
    >
      <div
        class="select-option__icon-container"
      >
        <svg>
          open.svg
        </svg>
      </div>
      <div
        class="select-option__label"
      >
        test 1
      </div>
      <div
        class="select-option__description"
      >
        another test 1
      </div>
      <input
        checked=""
        class="select-option__radio"
        type="radio"
      />
    </button>
    <button
      class="select-option select-option--1"
      tabindex="0"
    >
      <div
        class="select-option__icon-container"
      >
        <svg>
          lock-closed.svg
        </svg>
      </div>
      <div
        class="select-option__label"
      >
        test 2
      </div>
      <div
        class="select-option__description"
      >
        another test 2
      </div>
      <input
        class="select-option__radio"
        type="radio"
      />
    </button>
    <button
      class="select-option select-option--2"
      tabindex="0"
    >
      <div
        class="select-option__icon-container"
      >
        <svg>
          key-protected.svg
        </svg>
      </div>
      <div
        class="select-option__label"
      >
        test 3
      </div>
      <div
        class="select-option__description"
      >
        another test 3
      </div>
      <input
        class="select-option__radio"
        type="radio"
      />
    </button>
  </div>
</div>
`;
